components/experimental/src/units/ratio.rs

```diff
@@ -0,0 +1,228 @@
+// TODO: add test cases for IcuRatio.
+/// A ratio type that uses `BigInt` as the underlying type.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct IcuRatio(Ratio<BigInt>);
+
+#[derive(Debug)]
+pub enum IcuRatioError {
+    /// Represents an error when parsing a `BigInt` from a string.
+    /// For example, the string "1/23" is invalid because it contains a '/' character.
+    ///              the string "3A" is invalid because it contains a non-numeric character.
+    BigIntParseError(num_bigint::ParseBigIntError),
+
+    /// Represents an error when parsing a ratio from a string.
+    /// For example, the string "1/0/2" is invalid because it contains more than one '/' character.
+    InvalidRatioString,
+}
+
+impl IcuRatio {
+    /// Creates a new `IcuRatio` from the given numerator and denominator.
+    pub fn new(numerator: BigInt, denominator: BigInt) -> Self {
+        Self(Ratio::new(numerator, denominator))
+    }
+
+    /// Extracts the factor ratio from `ConversionInfoULE`.
+    pub fn factor_from_conversion_info(conversion_info: &ConversionInfoULE) -> Self {
+        let sign: num_bigint::Sign = Sign::from_unaligned(conversion_info.factor_sign).into();
+        Self::from(Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, conversion_info.factor_num().as_ule_slice()),
+            BigInt::from_bytes_le(
+                num_bigint::Sign::Plus,
+                conversion_info.factor_den().as_ule_slice(),
+            ),
+        ))
+    }
+
+    /// Extracts the offset ratio from `ConversionInfoULE`.
+    pub fn offset_from_conversion_info(conversion_info: &ConversionInfoULE) -> Self {
+        let sign: num_bigint::Sign = Sign::from_unaligned(conversion_info.offset_sign).into();
+        Self::from(Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, conversion_info.offset_num().as_ule_slice()),
+            BigInt::from_bytes_le(
+                num_bigint::Sign::Plus,
+                conversion_info.offset_den().as_ule_slice(),
+            ),
+        ))
+    }
+
+    /// Applys the given SI prefix to the ratio.
+    /// For example: if the ratio is 1/3 and the SI prefix is "kilo" (10^3),
+    ///              the result will be 1000/3.
+    /// Another example: if the ratio is 1/3 and the SI prefix is "kibi" ("2^10"),
+    ///                  the result will be 1024/3.
+    pub fn apply_si_prefix(&mut self, si_prefix: &SiPrefix) {
+        match si_prefix.base {
+            Base::Decimal => {
+                *self *= IcuRatio::ten().pow(si_prefix.power as i32);
+            }
+            Base::Binary => {
+                *self *= IcuRatio::two().pow(si_prefix.power as i32);
+            }
+        }
+    }
+
+    /// Returns the numerator of the ratio.
+    pub fn numerator(&self) -> &BigInt {
+        self.0.numer()
+    }
+
+    /// Returns the denominator of the ratio.
+    pub fn denominator(&self) -> &BigInt {
+        self.0.denom()
+    }
```

Comment:
Suggestion: since this is a public type, it would be better if we didn't publicly return `&BigInt` since then we must always be able to return a `&BigInt`. Once of the goals of IcuRatio was to make the internals private, so that we could eventually make this into a trait or change the underlying bigint library or something like that. Can you start with these being `pub(crate)` functions? If you need them public for some reason, then maybe make a `pub(crate)` version that returns a borrow and a public one named `numerator_bigint()` that returns `BigInt`.

Reply:
I removed them right now, I made all the functions that do not need public, to be pub(crate), and add what is needed when I am implementing FFI or use it in the datagen.